File: internal/providers/codex/reconcile.go
package codex

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/drogers0/aistat/v2/internal/accounts"
	"github.com/drogers0/aistat/v2/internal/cred"
)

// ReconcileInput is the full input set for Reconcile and ResolveActiveUUID.
//
// LookupID is called with the live id_token when no byte-match is found and
// the id_token is non-empty. It must not be nil when LiveBlob is non-nil.
// Identity resolution is pure (D1: JWT decode, no network).
type ReconcileInput struct {
	LiveBlob *cred.Credential                         // nil if absent; Raw holds exact live bytes
	Stored   []accounts.Account
	LookupID func(idToken string) (sub, email string, err error)
	Now      time.Time
}

// ReconcileOutput is the result of a Reconcile call.
type ReconcileOutput struct {
	Accounts     []accounts.Account
	ActiveUUID   string           // "" if none
	CaptureWarn  string           // non-empty when fallback applied (D3)
	Inserted     bool             // true if a new account slot was created
	Upserted     bool             // true if an existing slot was updated
	LiveUnstored *cred.Credential // non-nil when fallback applied; render-only
}

// Reconcile executes the Codex D-tree over the live credential and stored
// account slots. It is a pure function — all I/O (file read, JWT decode) is
// pre-resolved and passed via in. The caller is responsible for persisting
// out.Accounts when out.Inserted or out.Upserted is true.
func Reconcile(in ReconcileInput) ReconcileOutput {
	out := ReconcileOutput{
		Accounts: make([]accounts.Account, len(in.Stored)),
	}
	// Shallow copy: RawBlob is []byte (reference type); every code path that
	// changes RawBlob assigns a fresh slice — never index-writes in place.
	copy(out.Accounts, in.Stored)

	if in.LiveBlob == nil {
		return out
	}

	matchIdx, sub, email, lookupErr := findActive(in)

	switch {
	case matchIdx >= 0:
		// Byte-match: upsert RawBlob and LastSeenAt without a LookupID call.
		slot := out.Accounts[matchIdx]
		slot.RawBlob = json.RawMessage(in.LiveBlob.Raw)
		slot.LastSeenAt = in.Now
		out.Accounts[matchIdx] = slot
		out.ActiveUUID = slot.UUID
		out.Upserted = true

	case lookupErr != nil:
		// Identity lookup failed (missing id_token or parse error): render an
		// unstored live row and emit CaptureWarn (D3).
		out.LiveUnstored = in.LiveBlob
		out.CaptureWarn = fmt.Sprintf(
			"aistat: codex: could not identify live account (%s); rendering live row without storing — run `codex login` if this persists across runs",
			lookupErr.Error(),
		)

	default:
		// LookupID succeeded. UUID wins for matching; overwrite email if drifted.
		for i, acct := range out.Accounts {
			if acct.UUID == sub {
				slot := out.Accounts[i]
				slot.Email = email
				slot.RawBlob = json.RawMessage(in.LiveBlob.Raw)
				slot.LastSeenAt = in.Now
				out.Accounts[i] = slot
				out.ActiveUUID = sub
				out.Upserted = true
				return out
			}
		}
		// No UUID match — insert a new slot.
		out.Accounts = append(out.Accounts, accounts.Account{
			UUID:       sub,
			Email:      email,
			LastSeenAt: in.Now,
			RawBlob:    json.RawMessage(in.LiveBlob.Raw),
		})
		out.ActiveUUID = sub
		out.Inserted = true
	}

	return out
}

// ResolveActiveUUID is the read-only variant used by FetchForSwitch to identify
// the active account UUID without any store mutations. All LookupID failures
// (parse errors, missing id_token) return ("", nil) — not an error — because
// D1 guarantees there are no transient network failures on the identity path
// (pure JWT decode). T4's switch.go must not depend on surfaced errors from
// Codex ResolveActiveUUID.
func ResolveActiveUUID(in ReconcileInput) (string, error) {
	if in.LiveBlob == nil {
		return "", nil
	}

	matchIdx, sub, _, lookupErr := findActive(in)

	if matchIdx >= 0 {
		return in.Stored[matchIdx].UUID, nil
	}

	if lookupErr != nil {
		// All LookupID failures are parse errors (D1); never transient.
		return "", nil
	}

	return sub, nil
}

// findActive resolves the active account from the live blob without mutating
// in.Stored. It first scans for a byte-match; if none is found it calls
// LookupID via the id_token extracted from the live blob.
//
// Returns:
//   - matchIdx >= 0: index into in.Stored of the byte-matched slot. No LookupID call.
//   - matchIdx == -1, lookupErr == nil: LookupID succeeded; sub and email populated.
//   - matchIdx == -1, lookupErr != nil: byte-match missed and LookupID failed (or
//     id_token was absent/empty; LookupID is then never called).
//
// Precondition: in.LiveBlob is non-nil.
func findActive(in ReconcileInput) (matchIdx int, sub, email string, lookupErr error) {
	// Step 1: byte-match. First match wins (pathological duplicate-AT case).
	for i, acct := range in.Stored {
		if StoredAccessToken(acct) == in.LiveBlob.AccessToken {
			return i, "", "", nil
		}
	}

	// Step 2: extract id_token from live blob; guard before calling LookupID.
	idToken := extractIDToken(in.LiveBlob.Raw)
	if idToken == "" {
		return -1, "", "", fmt.Errorf("no id_token in live credential")
	}

	s, e, err := in.LookupID(idToken)
	if err != nil {
		return -1, "", "", err
	}
	return -1, s, e, nil
}

// extractIDToken parses tokens.id_token from raw Codex auth.json bytes.
// Returns "" on any error (missing field, malformed JSON, etc.).
func extractIDToken(raw []byte) string {
	var r struct {
		Tokens struct {
			IDToken string `json:"id_token"`
		} `json:"tokens"`
	}
	if err := json.Unmarshal(raw, &r); err != nil {
		return ""
	}
	return r.Tokens.IDToken
}

File: src/providers/codex/reconcile.rs
use super::account::stored_access_token_owned;
use crate::accounts::Account;
use crate::cred::{self, Credential};
use chrono::{DateTime, Utc};
use serde_json::value::RawValue;

pub struct ReconcileInput<'a> {
    pub live_blob: Option<&'a Credential>,
    pub stored: &'a [Account],
    pub now: DateTime<Utc>,
}

pub struct ReconcileOutput {
    pub accounts: Vec<Account>,
    pub active_uuid: String,
    pub capture_warn: Option<String>,
    pub inserted: bool,
    pub upserted: bool,
    pub live_unstored: Option<Credential>,
}

pub fn reconcile(input: ReconcileInput<'_>) -> ReconcileOutput {
    let live = match input.live_blob {
        Some(l) => l,
        None => {
            return ReconcileOutput {
                accounts: input.stored.to_vec(),
                active_uuid: String::new(),
                capture_warn: None,
                inserted: false,
                upserted: false,
                live_unstored: None,
            };
        }
    };

    // Step 1: byte-match
    for acct in input.stored {
        if stored_access_token_owned(acct) == live.access_token {
            let mut updated = acct.clone();
            updated.last_seen_at = input.now;
            if let Ok(blob) = RawValue::from_string(String::from_utf8_lossy(&live.raw).into_owned()) {
                updated.raw_blob = blob;
            }
            let mut accounts = input.stored.to_vec();
            for a in &mut accounts {
                if a.uuid == acct.uuid {
                    *a = updated.clone();
                }
            }
            return ReconcileOutput {
                accounts,
                active_uuid: acct.uuid.clone(),
                capture_warn: None,
                inserted: false,
                upserted: true,
                live_unstored: None,
            };
        }
    }

    // Step 2: decode id_token from live blob
    let id_token = cred::codex::extract_id_token(&live.raw);
    if let Some(ref token) = id_token {
        if let Ok((sub, email, _exp)) = cred::parse_codex_id_token(token) {
            let mut accounts = input.stored.to_vec();
            let existing = accounts.iter_mut().find(|a| a.uuid == sub);

            let raw_blob = RawValue::from_string(
                String::from_utf8_lossy(&live.raw).into_owned(),
            )
            .unwrap_or_else(|_| serde_json::value::RawValue::from_string("{}".into()).unwrap());

            let (inserted, upserted) = if let Some(acct) = existing {
                acct.email = email.clone();
                acct.last_seen_at = input.now;
                acct.raw_blob = raw_blob;
                (false, true)
            } else {
                accounts.push(Account {
                    uuid: sub.clone(),
                    email: email.clone(),
                    display_name: String::new(),
                    rate_limit_tier: String::new(),
                    last_seen_at: input.now,
                    raw_blob,
                });
                (true, false)
            };

            return ReconcileOutput {
                accounts,
                active_uuid: sub,
                capture_warn: None,
                inserted,
                upserted,
                live_unstored: None,
            };
        }
    }

    // Fallback
    ReconcileOutput {
        accounts: input.stored.to_vec(),
        active_uuid: String::new(),
        capture_warn: Some(
            "aistat: codex: could not identify live credential; rendering as unstored account".into(),
        ),
        inserted: false,
        upserted: false,
        live_unstored: Some(live.clone()),
    }
}
